src/3d/mesh/qgsmesh3dmaterial_p.cpp

```diff
@@ -141,14 +143,105 @@ class ColorRampTexture: public Qt3DRender::QAbstractTextureImage
     double mVerticalScale = 1;
 };
 
+class ArrowsGrid
+{
+  protected:
+    ArrowsGrid( const QVector<QgsVector> &vectors, const QSize &size ) :
+      mVectors( vectors ),
+      mSize( size )
+    {}
+    const QVector<QgsVector> mVectors;
+    const QSize mSize;
+
+    float length( int i ) const {return mVectors.at( i ).length();}
+    float angle( int i ) const {return mVectors.at( i ).angle();}
+};
+
+class ArrowsTextureGenerator: public ArrowsGrid, public Qt3DRender::QTextureImageDataGenerator
+{
+  public:
+    ArrowsTextureGenerator( const QVector<QgsVector> &vectors, const QSize &size ):
+      ArrowsGrid( vectors, size )
+    {}
+
+    Qt3DRender::QTextureImageDataPtr operator()() override
+    {
+      Qt3DRender::QTextureImageDataPtr dataPtr = Qt3DRender::QTextureImageDataPtr::create();
+      dataPtr->setFormat( QOpenGLTexture::RG32F );
+      dataPtr->setTarget( QOpenGLTexture::Target2D );
+      dataPtr->setPixelFormat( QOpenGLTexture::RG );
+      dataPtr->setPixelType( QOpenGLTexture::Float32 );
+
+      QByteArray data;
+
+      dataPtr->setWidth( mSize.width() );
+      dataPtr->setHeight( mSize.height() );
+      dataPtr->setDepth( 1 );
+      dataPtr->setFaces( 1 );
+      dataPtr->setLayers( 1 );
+      dataPtr->setMipLevels( 1 );
+
+
+      if ( mSize.width() > 0 && mSize.height() > 0 )
```

Comment:
It was to prevent executing the for loop if width=-1 and height=-1, if size is invalid, but I didn't remember QSize::isValid().
And sure, it is better to use isValid().